File: zig/grains/grains.zig
const math = @import("std").math;

pub const ChessboardError = error{
    IndexOutOfBounds,
};

pub fn square(index: usize) ChessboardError!u64 {
    if (index <= 0 or index > 64) return ChessboardError.IndexOutOfBounds;
    return math.pow(u64, 2, index - 1);
}

pub fn total() u64 {
    return @as(u64, 0) -% 1;
}

File: c/grains/grains.c
#include "grains.h"

uint64_t square(uint8_t index) {
    if (index <= 0 || index > 64) {
        return 0;
    }

    uint64_t ans = 1;
    for (uint8_t i = 0; i < index - 1; i++) {
        ans *= 2;
    }
    return ans;
}

uint64_t total() {
    uint64_t ans = 0;
    for (int i = 1; i <= 64; i++) {
        ans += square(i);
    }
    return ans;
}